### ISSUE TITLE
Very poor performance observed in powerflow/meter.cpp

### ISSUE THREAD

**Issue by dchassin** [gridlab-d/gridlab-d#1087]
There's a suspicious loop in powerflow/meter::postsync() that may be the source of the problem:

https://github.com/gridlab-d/gridlab-d/blob/eee842c92b5a810f597f3194759d776d49b613e2/powerflow/meter.cpp#L567 `last_measured_avg_voltage_mag[0] = last_measured_avg_voltage_mag[0] + ((last_measured_voltage[0].Mag() - last_measured_avg_voltage_mag[0])/(interval_dt + i));`

This can take a long time when dt is large (i.e., 3600).  I'm not even certain what this computation is really doing, so I'm not certain how to improve its performance.

**Comment by trevorhardy** [comment]
Is this in develop branch? 

It looks like something Andy added in for use in a specific project I'm working on. The project has an AMI requirement to report voltage values (such as the average value) at regular update intervals. For example, reporting the average voltage over the past 15 minutes (when the meter is also reporting energy consumption over that same interval).

**Comment by dchassin** [comment]
Yes. I tried removing the loops to test the hypothesis and it does run faster, but not as much as I had anticipated/hoped.  

I'm particularly interested/concerned with how the average is being calculated.  I'm not questioning whether it is needed.  Having a running average of voltage seems useful for a variety of reasons.

**Comment by dchassin** [comment]
Another thing I tried was to replace all the .Mag() with a single one that avoids repeating the same computation. That didn't help very much, except the ones inside the loop.

**Comment by dchassin** [comment]
One important point: if one of the meters is a swing bus, then the solver time gets counted as part of the meter performance profile.  So, while there are certainly some performance issues potentially with the voltage average calculations, the fact that the power flow solver time gets picked up by the meter when one of the meters is a swing bus can be a really big factor.  As it happens, that is the case for the some of the results I've seen.  I'll check with NG to see if that's the case with their system too.

**Comment by dchassin** [comment]
It might be worthwhile to look for a way to execute the solver "outside" the swing bus object's sync operation so as to not confuse the class's profiler.

**Comment by ftuffner** [comment]
I haven't had a chance to look at the forum post thoroughly, but my initial inclination is the comment about the swing bus above is 99% of the "disparity" in performance, especially between FBS and NR (at least for profiler results).

Regarding moving it outside of the node::sync function, I guess I have two fundamental questions: what do we intend the profiler to be used for and how accurate should it be?  While the NR-solver issue is the most prevalent example of this throwing the profiler off, other "cross-calling" objects could have the same issue.

I've always viewed the profiler as more of a very high-level analysis tool.  If we want it to be more accurate for performance evaluation, we may need to figure out how to capture things like function calls/external routines more effectively.

**Comment by dchassin** [comment]
I think you've nailed the issue there.  The profiler wasn't designed to handle that kind of analysis.  What's really needed is a separate profiling element for module-level functions, which node::sync technically is.  I would put this in the category of "consider the next time the profiler is upgraded".
